### ISSUE TITLE
possible deadlock when running DF for some time

### ISSUE THREAD

**Issue by romange** [dragonflydb/dragonfly#210]
as reported by @anosulchik

We have three instances r5.4xlarge, two of them were configured with 60% of memory available to dragonflydb and one with default settings (it gets 80%). The former two remained reachable so I was able to say that dragonflydb was just hanging on incoming requests, the latter instance became unavailable.

**Desktop (please complete the following information):**
* OS: Ubuntu 20.04.4 LTS
* Kernel: 5.13.0-1031-aws
* Containerized?: Bare Metal at EC2 under monit
* Version v0.5.0

**Comment by romange** [comment]
@anosulchik I opened this bug to discuss the issue. It's vaguer than the previous one, so I appreciate your help with that.  I have some questions regarding the problem:
 1. How does your app interact with these 3 instances? Does it send mirrored traffic (same commands) to all of them or each instance receives different commands?
 2. You said that instances A,B were reachable but C was not. Did you try to ssh into C? Can you usually ssh into these instances?
 3. Were all instances "hanging" but only C was unreachable? What does it mean, unreachable? did you try to telnet into it via 6379 port? Did you try to access it via browser? (DF supports http protocol via 6379).
 4. How did you conclude that DF was just "hanging on incoming requests"? Did you run "client list" on reachable instances?

I would also appreciate if you could send me "info all" output from C after it has been running for some time just to understand what type of commands you use there. Also I would also love to know what redis library do you use in order to access DF.

If you feel more comfortable talking offline, please send your answers to roman -at- dragonflydb.io

btw, I am still now sure if it's a deadlock in the backend or something else. For example, if you can usually ssh into your instances but you could not do it when it became unreachable, then it tells one story. But if you can ssh into the instance when it becomes "unreachable", then it tells absolutely different story.

**Comment by anosulchik** [comment]
@romange Please see replies below:

1. Writer app (side note: it's autoscaled app running at 15-30 instances) has three instances of dragonflydb in its configuration. The only command we use is `setex ...` which is sent to all three dragonflydb in best effort mode (write is considered as successful if at least one dragonflydb replied). At this stage there are no readers apps, only writes in our POC.

2. Please see screenshot for the reference: https://github.com/dragonflydb/dragonfly/issues/207#issuecomment-1185947981. Instances named `option` and `joel` were configured with explicit memory limit of `60%`, while `royal` was configured with default 80% (`--maxmemory` wasn't present in startup line for dragonflydb at that instance). When all three dragonflydb stopped processing traffic I was able to ssh to `option` and `joel` and connection to `royal` wasn't possible (it was timing out) so I had to restart `royal` to ssh into it. As for `option` and `joel` -- dragonflydb processes were in memory in `Interruptible sleep` state (`Sl` in `ps ax | grep dragonfly`) but did not processed any incoming requests -- application was throwing timeouts.

3. Just to confirm that unfortunately I didn't try to telnet to dragonflydb port but checked `netstat -lnp`-- it was there. Please see the screenshots - number of connected clients dropped from ~20 to 1-2 (`joel` and `option`, `royal` hanged completely and wasn't reporting anything). I believe those remaining clients are our redis exporters (it sits in memory and communicates to dragonflydb during prometheus scrape every other minute). Since there were metrics in place on the dashboard, I'm pretty sure that dragonflydb was responding to redis exporter but was not responding to setex commands.

4. I didn't run `client list` but I can try it next time. The graphs show that the problems with timeouts started as soon dragonflydb reached it's ceiling on allocated memory (60% and 80%) -- I think I can lower this number to something like 10% and reproduce the problem earlier (without waiting for ~4 hours).

Our app is written in Go and we're using https://github.com/gomodule/redigo as a client for dragonflydb since it's redis compatible.

I agree with you that it looks like a deadlock on application side since all three dragonflydb instances stopped processing traffic exactly at the same time (see number of processed commands dropped at the same time). But it also can be explained by our design - app is always writing to all three  dragonflydb instances, so after 4h of operation it could reach some bad state due to saturated memory at the same moment. 

I can re-run the POC -- feel free to let me know what checks, logs, probes you can offer to do to get it closer to find the root cause.

**Comment by romange** [comment]
Thanks Artem, this helps. You are saying that the only thing you did is SETEX, it really limits the problem space. The fact that all apps stopped at the same time is also a very interesting signal. Now, I looked at the graphs again and I have some bold theory. 
I am most likely mistaken, but you can verify it easily.

1. There is a bug in dragonfly that caused it to grow beyond its memory limit as you can see yourself ![here](https://user-images.githubusercontent.com/14876304/179311924-c1dafff8-5e42-47a3-b158-36efaed90d46.jpeg)

With Royal reaching 100% of memory capacity. 
2. Your instances are configured with swapping enabled.
3. Your writer app, even though writes in best effort mode, writes into each one of the servers sequentially, so if one of them is stuck, it does not continue to the next one. So best effort holds only if a connection is broken, or an error is returned. 
But in case of a stuck reply it does not proceed further.
4. Your royal server started a swapping procedure, and in this state almost any processing is halted,  no responses are returned.
5. All your apps are stuck on royal server but you think all servers are stuck. 

Let's start with the basics: 
1. Can you verify whether swappiness is active on dragonfly servers `cat /proc/swaps`
2. Can you verify how your app writes into servers. whether it really sends requests in parallel by issuing 3 independent requests and then proceeding even if some of them do not return. Note, it may issue 3 requests in parallel and then wait for all of them to return - it will also, IMHO, cause the stuck effect.

**Comment by anosulchik** [comment]
Thank you for your reply. Please see the answers to your questions:

1. Let me confirm that we don't use swap, there's no configured swap devices at any of dragonfly instances. We use ubuntu focal LTS EC2 image and I think it's disabled there by default.

2. The application performs *synchronized* parallels write into all redises/dragonfltdb. In other words, it sends three parallel requests to different redis connections and waits for response from all three before to proceed. It is literally the same code as shown in this example for sync package (we use it in our app) -- https://pkg.go.dev/sync#example-WaitGroup Please see `wg.Wait()` in the end of the method -- it waits until each batch of three setex either successfully finishes or fails (all three or just one). The app will mark the message as completed if at least one write into redis was successfull.

Please see timeout we use for our redis client:

```
const (
	connectTimeout  = 2
	readTimeout     = 2
	writeTimeout    = 2
	idleTimeout     = 60
)
```

There's no doubts there was memory management issue at `royal` and I agree there could be the case when hanged connection to one redis can pause the whole batch of three setex but we've got short timeout values so I'm pretty sure that the app did not stuck for longer than a few seconds (the graphs shows the outage lasted much much longer than timeout values).

**Comment by romange** [comment]
@anosulchik fair enough. Maybe the swappiness is not the root reason for Royal to be stuck but I still want to be sure thatwhat happens with `joel` and `option` is not affected by `royal`.  Regarding your remark with timeout values - it's true that it would progress after 2 seconds, but then it means you would have an effective command rate of `1/2s=0.5qps` all the time and that's not something you could see on the graphs you attached.

I suggest that you repeat the experiment with exactly the same setup.

1. try to ssh to royal **before** its memory reaches 100%, say after it crosses 85-90% but before it gets stuck.  
I also suggest that you also connect to the machine via redis-cli around this time , run `INFO ALL` and save the output.
Once you ssh, your ssh connection should not be kicked out if the machine is low on memory so you could use it later when the event starts.
2. Once your apps are stuck, I suggest that your try to connect to DF instances on all 3 machines via redis-cli.
3. if you preserved your ssh connection to royal and the DF there is unresponsive to redis-cli, it could be nice to see what happens on the machine itself. i.e. what `/var/log/syslog` shows, its CPU usage (via top) . DF has also `/tmp/dragonfly.INFO` logs that can have something or maybe your monit system may have logs. "lsof -p <pid>" can show you all the open files, 
`cat /proc/<pid>/status` can show its memory stats. 

4. For `option` and `joel` (and royal if possible of course) I would record:
 ```
 redis-cli INFO ALL > info_<backendname>`date +%s`.out
 ``` 
output a few times with delay 10s between runs just to see that their stats are progressing. you can also run:

```
redis-cli CLIENT LIST > clients_<backendname>`date +%s`.out
````
to see what's the status of client connections.
 
*I still think that `Royal` affects the interactions of your app with `option` and `joel`. I believe that it's highly unlikely that all backends are deadlocked at the same time, especially with a simple command like SETEX. *

**Comment by anosulchik** [comment]
CC @romange 

Just to confirm that I was able to reproduce the problem we were taking about.

1. In my most recent POC run I removed max memory settings for another host of three (`option` this time), so after ~4 hours it got memory exhausted that stopped the whole application from writing into redis / dragonflydb. I'm pretty sure that the app exhausted its redis connection pool -- even when timeouts are short, number of incoming messages to process and write into redis is so high, that it can bloat connection pool. Moreover, I noticed that `option` was `partially` available -- it responded with huge delays like 10-15s seconds. As soon as EC2 instance was restarted, traffic to `all three dragonflydb` instances was restored.

2. I also run another test with rather low memory limit for dragonflydb -- 10GB instead of ~100GB. It went well and there were no hiccups for multiple hours (memory limit wasn't exceeded and I saw steady rate of evictions). 

Bottom line: default memory limit doesn't work as expected so dragonflydb takes more than 80%. Other than that, all looks good.

We'll run dragonflydb one more time before using it in production. I'll communicate in case of any issues we notice.

**Comment by anosulchik** [comment]
I think we can close this issue for now.

**Comment by romange** [comment]
I am glad it's not a deadlock. Deadlock are the worst!
I will open another issue with memory limit problems. @anosulchik
